### DISCUSSION TITLE
Proposal: Give annotative names to Key and Value in Dictionary to improve code readability

### DISCUSSION THREAD

**Discussion by palazzo-train** [dotnet/csharplang#4422] [Language Ideas]
## Summary
Add new syntax in Dictionary declaration to annotate the key and value for readability

## Motivation
in C# struct, we can know clearly the purpose of a variable by it's name. For example,

````C#
public struct Book
{
    public string title;
    public string author;
}
````
Then, i know `b.title` is a type of string and it's referring to title.

However in C# dictionary, we can only specify the type

````C#
Dictionary<string,string> d
````

Can we make the code more readable such that the key of the dictionary is type of string and it is referring to title, and the value is type of string and it is referring to author? That means, other people can easily know that d["J.R.R. Tolkien"] is a wrong use of the dictionary when reading the code.

## Proposed Syntax

from 
````C#
Dictionary<string,string>
````

to
````C#
Dictionary<title@string,author@string>
````

The `title` and `author` are only annotation to the type string in the key and the type string in the value. It does not alter anything as of current Dictionary, but the annotation `title` and `author` will become some informative meta data used by IDE

**Comment by YairHalberstadt** [comment]
One way to do this might be 1 element tuples:

```csharp
Dictionary<(string title),(string author)>
```

Ideally 1 element tuples would be backed by the Actual element type, not ValueTuple, allowing you to freely cast from `Dictionary<string, string>` to `Dictionary<(string title),(string author)>`.
